pip/pip-294.md

```diff
@@ -0,0 +1,67 @@
+# Background knowledge
+
+There are mainly two `LoadManager` implementation in Pulsar broker: `ExtensibleLoadManager` and `ModularLoadManagerImpl`. `ModularLoadManagerImpl` is the default load manager, and `ExtensibleLoadManager` is a new load manager which is proposed after 3.0.0 version.
+
+## ModularLoadManagerImpl
+`ModularLoadManagerImpl` rely on zk to store and synchronize metadata about load, which pose greate pressure on zk, threatening the stability of system. Every broker will upload its `LocalBrokerData` to zk, and leader broker will retrieve all `LocalBrokerData` from zk ,generate all `BundleData` from each `LocalBrokerData`, and update all `BundleData` to zk. 
+
+## ExtensibleLoadManager
+`ExtensibleLoadManager` depends on system topics and table views for load balance metadata store and replication. Though not using zk to store and synchronize metadata about load, it is still necessary to control the number of bundles that need to be updated, for which there is a `loadBalancerMaxNumberOfBundlesInBundleLoadReport` configuration in `ExtensibleLoadManager` that select the top k bundles.
+
+# Motivation
+
+## ModularLoadManagerImpl
+As every bundle in the cluster corresponds to a zk node, it is common that there are thousands of zk nodes in a cluster, which results into thousands of read/update operations to zk. This will cause a lot of pressure on zk.
+
+**As All Load Shedding Algorithm pick bundles from top to bottom based on throughput/msgRate, bundles with low throughput/msgRate are rarely be selected for shedding. So there is no need to update these bundleData to zk frequently.**
+
+## ExtensibleLoadManager
+As the number of bundles and the throughput in the cluster changes dynamically, users can't select a reasonable number of bundles easily, while based on throughput, we can be sure that bundles with throughput less than 0.1M is useless for load balance no matter how the cluster change.
+
+We can also add this throughput-based bundle report size control in the ExtensibleLoadManager, working together with `loadBalancerMaxNumberOfBundlesInBundleLoadReport`.
```

Comment:
From the motivation, I learned that this proposal's goal is to limit the number of bundle reports written to the meta store or system topic because a huge number of bundle reports will bring heavy pressure on the meta store and system topic.

What I don't understand is why `loadBalancerMaxNumberOfBundlesInBundleLoadReport` is not enough. 
- We can limit the load report bundler number to 20 for each broker. If you have 10 brokers, and the leader will get 200 bundles to decide if we need to migrate any bundles. In the next round, each broker will report 20 bundles report based on current bundle stats. In this way, we can limit the max number of bundle reports written to the meta store or system topic
- For ExtensibleLoadManager, the number of bundle report issue has less impact on the load balance because we use non-persistent topic to store the bundle load report.
- Add throughput-based bundle report size control has the following issues.
  - Users need to learn another configuration to understand the load-balance behavior
  - The value is hard to configure, especially when the Pulsar cluster has low throughput in all brokers
  - Our goal is to help the leader to decide which broker is overloaded and which bundle needs to be unloaded. The TopK bundles play a key role to help the leader to make the decision.

Reply:
To configure a good value for `loadBalancerMaxNumberOfBundlesInBundleLoadReport` is not a easy way as the throughput and the number of bundles change dynamically.

For example, if the shedding algorithm need to pick bundles with 100M total, and the least throughput of bundle in the topK bundle set is greater than 100M, there will be no candidate bundle for unloading.
We can't determine how much bundles do we need for load balancing, which is a dynamic value. Both the throughput and the number of bundles change dynamically.

But, we can be sure that bundles with throughput less than 0.1Kb is useless for loadbalance no matter how the cluster change in most of cases. For clusters with extremely low throughput, we do not need to enable this feature, for which **i suggest to disable this feature by default**. 
WDYT, @codelipenghui

Reply:
We can provide a tool for users in admin to generate the throughput distribution of bundles in cluster.
For example:
![image](https://github.com/apache/pulsar/assets/52550727/5d91c4ec-71a9-4b20-9b84-1618dad9dc34)

With the throughput distribution of bundles, users can determine the configuration value easily.

Reply:
>if the shedding algorithm need to pick bundles with 100M for total, and the least throughput of bundle in the topK bundle set is greater than 100M, there will be no candidate bundle for unloading.

Can't we just pick the least throughput bundle to unload, even though it may be greater than 100MB? The goal of loadbalance is to make loadbalance fast and try to unload on big bundle instead of unloading 10 small/middle bundles

Reply:
1. What if the least throughput bundle serves 1GB instead of 100MB? we can't gurantee the safety of just pick the least throughput bundle. 
2. LoadSheddingStrategy implementations like `ThresholdShedder` will not pick the least throughput bundle to unload if its throughput is greater than the `minimumThroughputToOffload`.
3. I doubt that we should prefer unload a few super huge bundles instead of a appropriate amount of middle bundles. IMO, unload huge bundles will cause notifitable decline of throughtput, while unload some middle bundles will make the process smoother, and unload small bundles barely benefit to load balance.

Reply:
I can see many variations in selecting bundles in the load data.

Maybe we can generalize this option by introducing `loadBalanceBundleLoadDataReportStrategy` with the options of.

1. num of bundles
2. % of the total number of bundles or local broker's number of bundles
3. throughput-based
4. others.

IMO, top k is more predictable than the number of producers and consumer and their throughputs because the max number of bundles are relatively more static.